ledger_device_sdk/src/hmac.rs

```diff
@@ -0,0 +1,198 @@
+    fn new(key: &[u8]) -> Self;
+
+    /// Computes a HMAC in one line by providing the complete input as well as the
+    /// output buffer.
+    /// An error can be returned if one of the parameter is invalid
+    /// or if the output buffer size is not enough.
+    fn hmac(&mut self, input: &[u8], output: &mut [u8]) -> Result<(), HMACError> {
+        let err = unsafe {
+            cx_hmac_no_throw(
+                self.as_ctx_mut(),
+                CX_LAST,
+                input.as_ptr(),
+                input.len(),
+                output.as_mut_ptr(),
+                output.len(),
+            )
+        };
+        if err != CX_OK {
+            Err(err.into())
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Updates the current HMAC object state with the given input data.
+    /// This method may be called as many times needed (useful for large bufferized
+    /// inputs). This method should not be called after `finalize`.
+    /// An error can be returned if the input is invalid or the context in a wrong state.
+    fn update(&mut self, input: &[u8]) -> Result<(), HMACError> {
+        let err = unsafe { cx_hmac_update(self.as_ctx_mut(), input.as_ptr(), input.len()) };
+        if err != CX_OK {
+            Err(err.into())
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Finalizes the computation of the MAC and stores the result in the output buffer
+    /// as well as returning the MAC length.
+    /// This method should be called after one or many calls to `update`.
+    /// An error can be returned if one of the parameter is invalid
+    /// or if the output buffer size is not enough.
+    fn finalize(&mut self, output: &mut [u8]) -> Result<usize, HMACError> {
+        let mut out_len = output.len();
+        let err = unsafe { cx_hmac_final(self.as_ctx_mut(), output.as_mut_ptr(), &mut out_len) };
+        if err != CX_OK {
+            Err(err.into())
+        } else {
+            Ok(out_len)
+        }
+    }
+}
+
+/// This macro can be used to implement the HMACInit trait for a given hash
+/// algorithm by providing the structure name, the C context name, and the C
+/// context initialization function.
+macro_rules! impl_hmac {
+    ($typename:ident, $ctxname:ident, $initfname:ident) => {
+        #[derive(Default)]
+        #[allow(non_camel_case_types)]
+        pub struct $typename {
+            ctx: $ctxname,
+        }
+        impl HMACInit for $typename {
+            fn as_ctx_mut(&mut self) -> &mut cx_hmac_t {
+                unsafe { mem::transmute::<&mut $ctxname, &mut cx_hmac_t>(&mut self.ctx) }
+            }
+
+            fn as_ctx(&self) -> &cx_hmac_t {
+                unsafe { mem::transmute::<&$ctxname, &cx_hmac_t>(&self.ctx) }
+            }
+
+            fn new(key: &[u8]) -> Self {
+                let mut ctx: $typename = Default::default();
+                let _err = unsafe { $initfname(&mut ctx.ctx, key.as_ptr(), key.len() as u32) };
```

Comment:
@bboilot-ledger I am not sure to understand why two different versions of `impl_hmac` are required. In the C SDK, `no_throw` suffixed functions take a `size_t` as key length parameter whereas `throw` functions take an `unsigned int` key length parameter. On a 32-bit platform, both are 32 bits so passing key.len() (usize) should be ok ? Or am I missing anything 🤔 ?

Reply:
I completely agree with you. But the compiler is not really happy about that:
![image](https://github.com/user-attachments/assets/94cbc951-f54f-43d6-9def-100078d29291)
(same result if I try the opposite way).

I did try to check dynamically the function name and cast `key.len()` to the right type accordingly (which avoids code duplication) but still, it won't compile because the check is done at run time. I did not find a way to do "conditional" macros.
I may be missing something there (I am quite new on Rust), so if you see a way of avoiding code duplication, I will be super happy to implement it!

Reply:
Looks like using `try_into()` enables to avoid code duplication:
````
fn new(key: &[u8]) -> Self {
    let mut ctx: $typename = Default::default();
    let _err = unsafe {
        $initfname(&mut ctx.ctx, key.as_ptr(), key.len().try_into().unwrap())
    };
    ctx
}
````

Reply:
Oh nice, I did not think about that. Thanks for the help @yogh333 !